src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+- `neighbors`: the vector of the offsets (``o_i``) to the cartesian index of
+   the pixel (``p``), so that ``p + o_i`` is the cartesian index of the ``i``-th
+   neighbor of ``p``.
+
+# Details
+The pixels in the connected components generated by the method should be
+connected to each other by a path through neighboring pixels (as defined by
+`neighbors`).
+
+# See also
+[`MaxTree`](@ref)
+"""
+function rebuild!(maxtree::MaxTree{N},
+                  image::AbstractArray{<:Number, N},
+                  #=mask::AbstractArray{Bool, N},=#
+                  neighbors::AbstractVector{NTuple{N, Int}}) where N
+    # pixels need to be sorted according to their gray level.
+    sortperm!(maxtree.traverse, vec(image), rev=maxtree.rev)
+    # some sufficiently inner point
+    center = CartesianIndex{N}(ntuple(i -> 1 + size(image, i) ÷ 2, Val{N}()))
+    # offsets for linear indices (note that they are sorted in ascending order)
+    neighbor_offsets = getindex.(Ref(LinearIndices(image)),
+        [CartesianIndex{N}(ntuple(i -> center[i] + n[i], Val{N}())) for n in neighbors]) .-
+        getindex(LinearIndices(image), center)
+
+    # initialization of the image parent
+    parents = fill!(maxtree.parents, 0) |> vec
+    roots = fill!(similar(parents), 0) # temporary array containing current roots
+    rootpath = Vector{Int}() # temp array to hold the path to the current root within ancestors
+    cindexes = CartesianIndices(maxtree.parents)
+
+    # traverse the array in reversed order (from highest value to lowest value)
+    @inbounds for p in Iterators.Reverse(maxtree.traverse)
+        p_ci = cindexes[p]
+        parents[p] = roots[p] = p # it's a new root
+
+        for i in eachindex(neighbors)
+            # check that p is in the mask and that it's neighbor is a valid image pixel
+            (#=mask[p] && =#isvalid_offset(neighbors[i], p_ci, maxtree.axes)) || continue
+            index = p + neighbor_offsets[i] # linear index of the neighbor
+            (parents[index] == 0) && continue # ignore neighbor without parent (= its value is lower)
+            root = rootpath!(rootpath, roots, index) # neighbor's root
+            if (root != p) || (length(rootpath) > 1) # attach neighbor's root to the p
+                parents[root] = p
+                roots[rootpath] .= p # also attach the ancestral nodes of neighbor to p
+            end
+        end
+    end
+    return canonize!(maxtree, image)
+end
+
+# generates a vector of offsets to the CartesianIndex that defines
+# the neighborhood of the pixel in an N-dimensional image.
+# `connectivity` is the maximal number of orthogonal steps
+# (+1/-1 to a single dimension) that are required to reach a neighbor.
+function neighbor_cartesian_offsets(::Type{CartesianIndex{N}}, connectivity::Integer) where N
+    cis = CartesianIndices(ntuple(_ -> 3, Val{N}()))
+    offsets = [Tuple(ci) .- 2 for ci in cis]
+    return filter!(offset -> 0 < sum(abs, offset) <= connectivity, vec(offsets))
```

Comment:
Since you creates a new vector here `vec(offsets)`, it's a fake in-place operation

```suggestion
    return filter(offset -> 0 < sum(abs, offset) <= connectivity, offsets)
```

Reply:
`vec` creates a new vector object, but it shares the same data as the original `offset` array, so I'd rather keep as it is now.

Reply:
Oh, you are right here; haven't used `vec` before. Thanks for the explanation.